### ISSUE TITLE
Tensorflow uses only one from eight gpus

### ISSUE THREAD

**Issue by Lxnus** [tensorflow/tensorflow#23922]
Hey,
i have written an neural chatbot with tensorflow. But my problem is, that TF only uses one GPU.
I am training on an NVIDIA DGX Station with 8x Tesla V100, so I want to use alle memory and cores.
Where is my mistake in the code? Who could I solve my problem?

I would be very thankful about an answer! :D

My code:

import time

import data
import click
import numpy as np
import tensorflow as tf
import tensorlayer as tl

from tqdm import tqdm
from sklearn.utils import shuffle
from tensorlayer.layers import DenseLayer, EmbeddingInputlayer, Seq2Seq, retrieve_seq_length_op2

sessionConfig = tf.ConfigProto(allow_soft_placement = True, log_device_placement = False)

@click.command()
@click.option('-dc', '--data-corpus', default = 'twitter')
@click.option('-bs', '--batch-size', default = 32)
@click.option('-e', '--epochs', default = 100000)
@click.option('-lr', '--learning-rate', default = 0.001)
@click.option('-m', '--mode', is_flag = True)
def train(data_corpus, batch_size, epochs, learning_rate, mode):

    metadata, trainX, trainY, testX, testY, validX, validY = setup(data_corpus)

    srcLength = len(trainX)
    targetLength = len(trainY)

    assert srcLength == targetLength

    nStep = srcLength // batch_size
    srcVocabSize = len(metadata['index2Word'])
    embeddingDim = 1024

    word2Index = metadata['word2Index']  
    index2Word = metadata['index2Word']  

    unk_id = word2Index['unk']  
    pad_id = word2Index['_']    

    start_id = srcVocabSize  
    end_id = srcVocabSize + 1 

    word2Index.update({'start_id': start_id})
    word2Index.update({'end_id': end_id})
    index2Word = index2Word + ['start_id', 'end_id']

    srcVocabSize = tgt_vocab_size = srcVocabSize + 2

    target_seqs = tl.prepro.sequences_add_end_id([trainY[100]], end_id=end_id)[0]
    decode_seqs = tl.prepro.sequences_add_start_id([trainY[100]], start_id=start_id, remove_last=False)[0]
    target_mask = tl.prepro.sequences_get_mask([target_seqs])[0]
    if not mode:
        print("encode_seqs", [index2Word[id] for id in trainX[100]])
        print("target_seqs", [index2Word[id] for id in target_seqs])
        print("decode_seqs", [index2Word[id] for id in decode_seqs])
        print("target_mask", target_mask)
        print(len(target_seqs), len(decode_seqs), len(target_mask))
		
    tf.reset_default_graph()
    
    session = tf.Session(config=sessionConfig)
	
    encode_seqs = tf.placeholder(dtype=tf.int64, shape=[batch_size, None], name="encode_seqs")
    decode_seqs = tf.placeholder(dtype=tf.int64, shape=[batch_size, None], name="decode_seqs")
    target_seqs = tf.placeholder(dtype=tf.int64, shape=[batch_size, None], name="target_seqs")
    target_mask = tf.placeholder(dtype=tf.int64, shape=[batch_size, None], name="target_mask") 
        
    outputNetwork, _ = createModel(encode_seqs, decode_seqs, srcVocabSize, embeddingDim, isTrain=True, reuse=False)
    outputNetwork.print_params(False)
        
    encode_seqs2 = tf.placeholder(dtype=tf.int64, shape=[1, None], name="encode_seqs")
    decode_seqs2 = tf.placeholder(dtype=tf.int64, shape=[1, None], name="decode_seqs")
        
    network, RNNnetwork = createModel(encode_seqs2, decode_seqs2, srcVocabSize, embeddingDim, isTrain=False, reuse=True)
    y = tf.nn.softmax(network.outputs)
        
    loss = tl.cost.cross_entropy_seq_with_mask(logits=outputNetwork.outputs, target_seqs=target_seqs, input_mask=target_mask, return_details=False, name='cost')
        
    trainOp = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(loss)

    session.run(tf.global_variables_initializer())

    tl.files.load_and_assign_npz(sess=session, name='model.npz', network=network)	

    def inference(seed):
        seed_id = [word2Index.get(w, unk_id) for w in seed.split(" ")]
        
        state = session.run(RNNnetwork.final_state_encode, {encode_seqs2: [seed_id]})
        o, state = session.run([y, RNNnetwork.final_state_decode], {RNNnetwork.initial_state_decode: state, decode_seqs2: [[start_id]]})
        w_id = tl.nlp.sample_top(o[0], top_k=3)
        w = index2Word[w_id]
        sentence = [w]
        for _ in range(200): # 30 output length
            o, state = session.run([y, RNNnetwork.final_state_decode], {RNNnetwork.initial_state_decode: state, decode_seqs2: [[w_id]]})
            w_id = tl.nlp.sample_top(o[0], top_k=2)
            w = index2Word[w_id]
            if w_id == end_id:
                break
            sentence = sentence + [w]
        return sentence

    if mode:
        while True:
            input_seq = input('Input >> ')
            sentence = inference(input_seq)
            print("Output >>", ' '.join(sentence))
    else:
        seeds = ["happy birthday have a nice day",
                 "donald trump won last nights presidential debate according to snap online polls"]
        for epoch in range(epochs):
            trainX, trainY = shuffle(trainX, trainY, random_state=0)
            total_loss, n_iter = 0, 0
            for X, Y in tqdm(tl.iterate.minibatches(inputs=trainX, targets=trainY, batch_size=batch_size, shuffle=False), total=nStep, desc='Epoch[{}/{}]'.format(epoch + 1, epochs), leave=False):
                X = tl.prepro.pad_sequences(X)
                _target_seqs = tl.prepro.sequences_add_end_id(Y, end_id=end_id)
                _target_seqs = tl.prepro.pad_sequences(_target_seqs)
                _decode_seqs = tl.prepro.sequences_add_start_id(Y, start_id=start_id, remove_last=False)
                _decode_seqs = tl.prepro.pad_sequences(_decode_seqs)
                _target_mask = tl.prepro.sequences_get_mask(_target_seqs)
				
                _, loss_iter = session.run([trainOp, loss], {encode_seqs: X, decode_seqs: _decode_seqs, target_seqs: _target_seqs, target_mask: _target_mask})
                total_loss += loss_iter
                n_iter += 1

            print('Epoch [{}/{}]: loss {:.4f}'.format(epoch + 1, epochs, total_loss / n_iter))
            
            for seed in seeds:
                print("Input >>", seed)
                for _ in range(5):
                    sentence = inference(seed)
                    print("Output >>", ' '.join(sentence))
            
            tl.files.save_npz(network.all_params, name='model.npz', sess=session)
    
    session.close()
    
    
def createModel(encodeSeqs, decodeSeqs, vocabSize, embeddingDim, isTrain = True, reuse = False):
    with tf.variable_scope("model", reuse = reuse):
        with tf.variable_scope("embedding") as vs:
            networkworkEncoder = EmbeddingInputlayer(inputs = encodeSeqs, 
                                                 vocabulary_size = vocabSize, 
                                                 embedding_size = embeddingDim, 
                                                 name = 'sequenceEmbedding')
            vs.reuse_variables()
            networkworkDecoder = EmbeddingInputlayer(inputs = decodeSeqs, 
                                                 vocabulary_size = vocabSize, 
                                                 embedding_size = embeddingDim, 
                                                 name = 'sequenceEmbedding')
        networkworkRNN = Seq2Seq(networkworkEncoder, networkworkDecoder, 
                             cell_fn = tf.nn.rnn_cell.LSTMCell, 
                             n_hidden = embeddingDim, 
                             initializer = tf.random_uniform_initializer(-0.1, 0.1),
                             encode_sequence_length = retrieve_seq_length_op2(encodeSeqs),
                             decode_sequence_length = retrieve_seq_length_op2(decodeSeqs),
                             initial_state_encode = None,
                             dropout = (0.5 if isTrain else None),
                             n_layer = 3,
                             return_seq_2d = True,
                             name = 'seq2seq')
        networkworkOut = DenseLayer(networkworkRNN, n_units = vocabSize, act = tf.identity, name = 'output')
    return networkworkOut, networkworkRNN

def setup(data_corpus):
    metaData, indexQ, indexA = data.load_data(Path=''.format(data_corpus))
    (trainX, trainY), (testX, testY), (validX, validY) = data.split_dataset(indexQ, indexA)
    trainX = tl.prepro.remove_pad_sequences(trainX.tolist())
    trainY = tl.prepro.remove_pad_sequences(trainY.tolist())
    testX = tl.prepro.remove_pad_sequences(testX.tolist())
    testY = tl.prepro.remove_pad_sequences(testY.tolist())
    validX = tl.prepro.remove_pad_sequences(validX.tolist())
    validY = tl.prepro.remove_pad_sequences(validY.tolist())
    return metaData, trainX, trainY, testX, testY, validX, validY

def main():
    try:
        train()
    except KeyboardInterrupt:
        print('Aborted!')

if __name__ == '__main__':
    main()

**Comment by byronyi** [comment]
Please consider using DistributionStrategy API for transparent multi-GPU computation. Currently TF does not distribute the computation without any modification to your code.

This question is better asked on StackOverflow.

**Comment by ymodak** [comment]
Closing since solution is provided by @byronyi . Feel free to open a new issue if you are running into problems while using DistributionStrategy AP. Thanks!

**Comment by ymodak** [comment]
Are you satisfied with the resolution of your issue?
[Yes](https://goo.gl/forms/Oe0tEvODFRoI2gJF3)
[No](https://goo.gl/forms/fUjzOfrtkFbrOT8d2)
